### ISSUE TITLE
Add Auto start tools for running App, Session, Scripts

### ISSUE THREAD

**Issue by AC100Addict** [CypherpunkArmory/UserLAnd#368]
Dear all,

First, I would like to thank developers for their great works on this very useful and fabulous tool !

I am using UserLAnd with my Android Nvidia Shield TV with Oreo 8.0. (I side loaded it from an APK as it is not available on Android TV market yet.)

It works but I would like to automatically after boot:
1. Start UserLAnd
2. Start a session (vnc)
3. Run a script to export DISPLAY to a X server on my network
4. Keep UserLAnd running full speed in the background

My goal:
I would like to use the horse power of my shield on a weak/old arm smartbook which will run the X server only. My arm smartbook is sadly barely capable to handle more that the X Server without running out of RAM but its lightness, screen and keyboard make it perfect as an extension of my Shield TV for surfing or doing some office works.

**1. Will it be possible to have a setting to automatically start UserLAnd on android TV boot? And even restart it if it stops?**
Currently I use "Auto Start" app to launch it on boot.

**2. Will it be possible to select a default session to load automatically at start?**
Currently I have to manually click to select either an Application or a Session to start

**3. How can we create new App in UserLAnd to manager a customer script at launch?
Is there any wiki, tutorial or instruction to create custom UserLAnd Application ?**
This topic has been already discussed #56 , #60 and closed thanks to the App feature introduction, but I didn't find out how I can do my own script.
The latest solution would be to add a boot script within linux boot like #287 but I don't find it clean. Indeed, I would like also to have other sessions which open the same file system but with other scripts at boot.

**4. Will it be possible to prevent Android to slowdown or to stop UserLAnd from running in the background?**
So I can use it remotely (using the export DISPLAY function) while my girlfriend can still use the Shield TV for watching her series.

Sorry for my long post and my multiple questions.
I am very grateful for your precious help and idea to solve this out.

I am very enthusiastic about UserLAnd!
My case is simply showing one of the crazy combination allowed by running linux on no root android devices and which will leverage a lot from adding some autostart / restart / session launch / script running tools in UserLAnd.

**Comment by MatthewTighe** [comment]
To address some points:

1. I don't know enough about Android T.V. to say either way. We can look into it.

2. This is something we can look into in the future, but may not be possible while we rely on 3rd party applications for connecting to the servers starting on your device.

3. You can add scripts to run on startup to your filesystem like you would on any Linux device.

4. We're already doing everything we can (as far as I know) to prevent Android from killing UserLAnd.

**Comment by markstinson** [comment]
I second the request for Android TV (as I've a Shield TV as well) if possible.

I've noticed with the HAL launcher, there's Handheld, Handheld+Tablet, and Android TV hardware categories -- so it's very likely a 2nd build aimed for Android TV would be required. I don't know if you could get away with using an existing Landscape Material Design for the Big Screen.

It would be nice to use this, turning my Android TV into a more usable/functional workstation. There's several other Android TV brands/models out there - current and aging - that could benefit from this (to re-purpose them.

You might consider looking into https://github.com/termux as Termux has an available AndroidTV version in the Play Store as well. (The Termux accessories do not appear though - but are available via F-droid for a more convenient side-load.)

**Comment by hugorosario** [comment]
I am also interested in this features and plan on running a small home server (apache, php and mysql) on my android TV box and UserLAnd seems like a good bet.
Since I also have some experience coding for Android, I forked this repository and started working on this features myself.

I started by adding a new setting to the sessions which is "Start on boot".
My general idea is to listen for the BOOT_COMPLETED broadcast and start the UserLAnd service with an new action called "autostart" which will silently start all the sessions that have the "Start on boot" flag.
After starting the session, it will also execute a startup script which will be located somewhere (maybe on /support/autostart.sh or something like that). 

Will let you know when I have something useful...

**Comment by hugorosario** [comment]
Just to let you know.
I have almost completed the features I require.
I only need to make a few more tests and I might make a pull request.

In the meantime, if you have the Android Studio, you can checkout my repository and test this yourself.
The modified fork is here : https://github.com/hugorosario/UserLAnd

**Comment by brianddk** [comment]
> 3\. You can add scripts to run on startup to your filesystem like you would on any Linux device.

@MatthewTighe , can you give some hints on where do to this on a Debian filesystem?  It doesn't look like a lot of the standard `init` stuff is populated, probably due to being a chroot.  I can't seem to get the normal `init.d` stuff working.

**Comment by MatthewTighe** [comment]
@brianddk Yeah turns out I may have been wrong about that actually, since there's not actually ever a real device boot. One of the workarounds we're incorporating is to instead place scripts under `/etc/profile.d`. Scripts there are executed in alphabetical order upon login, which happens automatically with our flow since we ssh to connect to a running server as the session starts.

E.G. we run scripts that handle startup for the "apps" by placing them there with a name with the lowest alphabetical importance. You might conflict with the one we use for Debian if you're accessing your filesystem through the "apps" page, so just name it with a ton of `z`s.

**Comment by brianddk** [comment]
OK, I tried putting a script `/etc/profile.d/z_linuxrc` in there and it never seemed to execute.  As a hack, I just called the script from `/support/startSSHServer.sh`.  Not elegant, but its working.  I'll see if I can't log an issue on the `/etc/profile.d` stuff.

BTW... Love the product.  I'm hosting a Tor Hidden Service (`.onion`) on my phone.  Very cool since it manages to punch through firewalls as part of the onion protocol.

**Comment by MatthewTighe** [comment]
@brianddk Are you running your filesystem from the "apps" page? The ASCII value of `_` is less than `z`, so it would be called before the script we call on startup.

**Comment by brianddk** [comment]
> Are you running your filesystem from the "apps" page? 

@MatthewTighe , I'm running it from the sessions page.  It's just a Debian instance with ssh enabled.  I can try from the apps page as well as prepending an underscore.

**Comment by brianddk** [comment]
> Are you running your filesystem from the "apps" page?

@MatthewTighe , I tore down my sessions and ran from the "apps" page as you suggested and can now confirm that `profile.d` is working as you suggested.  Thx.

**Comment by rhlobo** [comment]
+1 for start on boot

@hugorosario made a pull request and I tried to finish the work needed, incouding translations. Made a PR for the PR. Master branch git ahead and I was asked to rewrite everytbing from scratch.

Unfortunately, I dont have the needed skills for that (and time to get up to that). My initial aditions already required a huge effort once I strugled in every aspect, cause had no pior knowledge on android, the IDE, kotlin, the project etc.

Please, keep this on the backlog. The pull requests that where made may serve as reference to reduce time needed by the paid team.

**Comment by MatthewTighe** [comment]
This will definitely stay on the backlog. Thanks for the time and effort you committed to try and make it happen!

**Comment by virtimus** [comment]
Hi there - used @hugorosario work to reapply the changes on master (on my fork) - for me works.
Use it if You want.

**Comment by omniotorr** [comment]
> @MatthewTighe I tried putting a script `/etc/profile.d/z_linuxrc` in there and it never seemed to execute. 

Because your "z_linuxrc" script doesn't have a ".sh" extension (this goes for most (all?) linux distros, it's not particular to userland). Rename the script to "z_linuxrc.sh" and it should be executed.

**Comment by corbinlc** [comment]
Closing as part of a repository cleanup. If this issue is still relevant, feel free to reopen it or file a fresh one.
